File: backend/app/models/datasets.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


DatasetStatus = Literal["processing", "ready", "failed"]
FileKind = Literal["dicom", "nifti", "image", "unknown"]


class DatasetIngestRequest(BaseModel):
    url: str
    name: str | None = None
    team_id: str | None = None


class DatasetSummary(BaseModel):
    total_files: int = 0
    scheduled_files: int = 0
    modality_counts: dict[str, int] = Field(default_factory=dict)
    modalities: dict[str, dict] = Field(default_factory=dict)
    mixed_modality: bool | None = None
    outliers: int = 0
    kind_counts: dict[str, int] = Field(default_factory=dict)
    ext_counts: dict[str, int] = Field(default_factory=dict)
    scheduled_ext_counts: dict[str, int] = Field(default_factory=dict)
    duplicate_basename_count: int = 0
    duplicate_basename_ext_counts: dict[str, int] = Field(default_factory=dict)
    image_2d_count: int = 0
    volume_3d_count: int = 0


class DatasetDoc(BaseModel):
    id: str = Field(alias="_id")
    name: str
    source_url: str
    team_id: str | None = None
    owner_user_id: str | None = None
    status: DatasetStatus
    created_at: datetime
    summary: DatasetSummary = Field(default_factory=DatasetSummary)
    meta: dict[str, Any] = Field(default_factory=dict)


class DatasetIngestResponse(BaseModel):
    dataset_id: str
    status: DatasetStatus
    # For notebook URLs that resolve to multiple datasets
    all_dataset_ids: list[str] | None = None
    source_type: str = "direct"  # "direct", "kaggle_notebook", etc.
    resolved_urls: list[str] | None = None


class DatasetListItem(BaseModel):
    dataset_id: str
    name: str
    status: DatasetStatus
    created_at: datetime
    summary: DatasetSummary


class DatasetSummaryResponse(DatasetSummary):
    stage: str = "unknown"
    modality_percentages: dict[str, float] = Field(default_factory=dict)


class FileListItem(BaseModel):
    dataset_id: str
    relpath: str
    kind: FileKind
    modality: str
    modality_model: dict = Field(default_factory=dict)
    ndim: int | None = None
    dims: list[int] | None = None
    size_bytes: int = 0
    created_at: datetime
    meta: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/api/datasets.ts
// API base - nginx proxies /api/* to backend
const API_BASE = '/api';

export interface DatasetSummary {
  total_files: number;
  scheduled_files: number;
  modality_counts: Record<string, number>;
  modalities: Record<string, Record<string, unknown>>;
  mixed_modality: boolean | null;
  outliers: number;
  kind_counts: Record<string, number>;
  ext_counts: Record<string, number>;
  scheduled_ext_counts: Record<string, number>;
  duplicate_basename_count: number;
  duplicate_basename_ext_counts: Record<string, number>;
  image_2d_count: number;
  volume_3d_count: number;
}

export interface DatasetListItem {
  dataset_id: string;
  name: string;
  status: 'processing' | 'ready' | 'failed';
  created_at: string;
  summary: DatasetSummary;
}

export interface DatasetSummaryResponse extends DatasetSummary {
  stage: string;
  modality_percentages: Record<string, number>;
}

export interface FileListItem {
  dataset_id: string;
  relpath: string;
  kind: 'dicom' | 'nifti' | 'image' | 'unknown';
  modality: string;
  modality_model: Record<string, unknown>;
  ndim: number | null;
  dims: number[] | null;
  size_bytes: number;
  created_at: string;
  meta: Record<string, unknown>;
}

export interface DatasetDetail {
  _id: string;
  name: string;
  source_url: string;
  team_id: string | null;
  owner_user_id: string | null;
  status: 'processing' | 'ready' | 'failed';
  created_at: string;
  summary: DatasetSummary;
  meta: Record<string, unknown>;
}

export interface IngestRequest {
  url: string;
  name?: string;
  team_id?: string;
}

export interface IngestResponse {
  dataset_id: string;
  status: 'processing' | 'ready' | 'failed';
  all_dataset_ids: string[] | null;
  source_type: string;
  resolved_urls: string[] | null;
}

export async function fetchDatasets(): Promise<DatasetListItem[]> {
  const res = await fetch(`${API_BASE}/datasets`);
  if (!res.ok) throw new Error('Failed to fetch datasets');
  return res.json();
}

export async function fetchDataset(datasetId: string): Promise<DatasetDetail> {
  const res = await fetch(`${API_BASE}/datasets/${datasetId}`);
  if (!res.ok) throw new Error('Failed to fetch dataset');
  return res.json();
}

export async function fetchDatasetSummary(datasetId: string): Promise<DatasetSummaryResponse> {
  const res = await fetch(`${API_BASE}/datasets/${datasetId}/summary`);
  if (!res.ok) throw new Error('Failed to fetch dataset summary');
  return res.json();
}

export async function fetchDatasetFiles(
  datasetId: string,
  skip = 0,
  limit = 200
): Promise<FileListItem[]> {
  const res = await fetch(`${API_BASE}/datasets/${datasetId}/files?skip=${skip}&limit=${limit}`);
  if (!res.ok) throw new Error('Failed to fetch dataset files');
  return res.json();
}

export async function ingestDataset(request: IngestRequest): Promise<IngestResponse> {
  const res = await fetch(`${API_BASE}/datasets/ingest`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  });
  if (!res.ok) throw new Error('Failed to ingest dataset');
  return res.json();
}
